maskit/ensembles.py

```diff
@@ -0,0 +1,179 @@
+        #   contain values from dropped gates
+        return EnsembleResult(
+            branch=selected_branch,
+            branch_name=branch_name,
+            cost=branch_costs[minimum_index].unwrap(),
+            gradient=branch_gradients[minimum_index],
+            brutto_steps=1 + len(branches) * ensemble_steps,
+            netto_steps=1 + ensemble_steps,
+            brutto=(
+                basic_active_gates
+                + sum(
+                    [branch.active() * ensemble_steps for branch in branches.values()]
+                )
+            ).unwrap(),
+            netto=(
+                basic_active_gates + selected_branch.active() * ensemble_steps
+            ).unwrap(),
+            ensemble=True,
+        )
+
+
+class IntervalEnsemble(Ensemble):
+    __slots__ = ("_interval", "_counter")
+
+    def __init__(self, dropout: Optional[Dict], interval: int):
+        super().__init__(dropout)
+        if interval < 1:
+            raise ValueError(
+                f"Interval must be in a valid range (>= 1), current value {interval}"
+            )
+        self._interval = interval
+        self._counter = 0
+        self.perturb = False
+
+    def _check_interval(self):
+        if self._counter % self._interval == 0:
+            self.perturb = True
+            self._counter = 0  # reset the counter
+        else:
+            self.perturb = False
+
+    def step(
+        self, masked_circuit: MaskedCircuit, optimizer, *args, ensemble_steps: int = 1
+    ) -> EnsembleResult:
+        self._counter += 1
+        self._check_interval()
+        return super().step(
+            masked_circuit, optimizer, *args, ensemble_steps=ensemble_steps
+        )
+
+
+class AdaptiveEnsemble(Ensemble):
+    __slots__ = ("_cost", "epsilon")
+
+    def __init__(
+        self,
+        dropout: Optional[Dict[str, Dict]],
+        size: int,
+        epsilon: float,
+    ):
+        if size <= 0:
+            raise ValueError(f"Size must be bigger than 0 (received {size})")
+        super().__init__(dropout)
+        self._cost = deque(maxlen=size)
+        self.epsilon = epsilon
+        self.perturb = False
+
+    def _check_cost(self, current_cost):
+        self.perturb = False
+        self._cost.append(current_cost)
+        if self._cost.maxlen and len(self._cost) >= self._cost.maxlen:
+            if current_cost > 0.1:  # evaluate current cost
+                if np.sum(np.diff(self._cost)) < self.epsilon:
+                    if __debug__:
+                        print("======== allowing to perturb =========")
+                    self._cost.clear()
+                    self.perturb = True
+
+    def step(
+        self, masked_circuit: MaskedCircuit, optimizer, *args, ensemble_steps: int = 1
+    ) -> EnsembleResult:
+        result = super().step(
+            masked_circuit, optimizer, *args, ensemble_steps=ensemble_steps
+        )
+        self._check_cost(result.cost)
```

Comment:
Is this supposed to be after the step? ``IntervalEnsemble`` does the corresponding thing before the step.

Reply:
Thanks, looks much cleaner when both act similarly :)